src/agent/application.cpp

```diff
@@ -61,20 +61,34 @@ Application::Application(const std::string               &aInterfaceName,
     , mBackboneInterfaceName(aBackboneInterfaceNames.empty() ? "" : aBackboneInterfaceNames.front())
 #endif
     , mNcp(mInterfaceName.c_str(), aRadioUrls, mBackboneInterfaceName, /* aDryRun */ false, aEnableAutoAttach)
+#if OTBR_ENABLE_MDNS_AVAHI || OTBR_ENABLE_MDNS_MDNSSD
+    , mPublisher(Mdns::Publisher::Create([this](Mdns::Publisher::State aState) { this->HandleMdnsState(aState); }))
+#else
+    , mPublisher(nullptr)
```

Comment:
Can we consider not having a `mPublisher` variable (and `GetPublisher()`) in this case? (when `OTBR_MDNS=OFF` is used).

- It would safer as if any other module uses `GetPublisher()` we know we have a valid (non-null) one.
- If build configs are incorrect, such that `OTBR_MDNS=OFF` but some other feature that uses the `Publisher` is enabled, we get a build-time error (i.e., missing `Application::GetPublisher()`)

---- 

Another  idea:
We can define a macro `OTBR_ENABLE_MDSN` which we define as:
```
#define OTBR_ENABLE_MDSN (OTBR_ENABLE_MDNS_AVAHI || OTBR_ENABLE_MDNS_MDNSSD)
```

Reply:
>We can define a macro OTBR_ENABLE_MDSN which we define as:

I also thought about this one. This PR seems a good chance to introduce it. It's added.